### DISCUSSION TITLE
Using ggml as a dependency, how can I prevent ggml.dll from getting generated on Windows builds?

### DISCUSSION THREAD

**Discussion by balisujohn** [ggml-org/ggml#881] [General]
here's my current CMakeLists.txt:
````
cmake_minimum_required(VERSION 3.12)
project("tortoise")


set(CMAKE_EXPORT_COMPILE_COMMANDS ON)

add_executable(tortoise main.cpp common.cpp)

option(GGML_CUBLAS "cuda mode" OFF)

if (GGML_CUBLAS)
    add_definitions(-DGGML_USE_CUBLAS)
endif()

add_subdirectory(ggml)

add_definitions(-DGGML_MAX_NAME=128)


set(GGML_STATIC  ON)



set_target_properties(tortoise PROPERTIES ARCHIVE_OUTPUT_DIRECTORY $<1:${PROJECT_SOURCE_DIR}/build>)
set_target_properties(tortoise PROPERTIES LIBRARY_OUTPUT_DIRECTORY $<1:${PROJECT_SOURCE_DIR}/build>)
set_target_properties(tortoise PROPERTIES RUNTIME_OUTPUT_DIRECTORY $<1:${PROJECT_SOURCE_DIR}/build>)

target_link_libraries(tortoise PUBLIC ggml)
target_include_directories(tortoise PUBLIC . )
target_compile_features(tortoise PUBLIC cxx_std_11)
````

Right now, I get `build/tortoise.exe` and `build/bin/Debug/ggml.dll`, and have to move the dll to the build folder to get the exe to work. What I'd like instead is to just get an exe and not generate the dll at all. I tried adding the line

````
set(GGML_STATIC  ON)
````

But it seems to have no effect.

**Comment by slaren** [comment]
You can disable building ggml as shared library by setting `BUILD_SHARED_LIBS` to `OFF`.

**Reply by balisujohn** [reply]
I added this line:
````
set(BUILD_SHARED_LIBS OFF)
````

and this is the compilation output:
````
MSBuild version 17.10.4+10fbfbf2e for .NET Framework

  1>Checking Build System
  Building Custom Rule C:/Users/John/tortoise_2/tortoise.cpp/ggml/src/CMakeLists.txt
  ggml.c
  ggml-alloc.c
  ggml-backend.c
  ggml-quants.c
  Generating Code...
C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\ucrt\assert.h(21,9): warning C4005: 'static_assert': macro
redefinition [C:\Users\John\tortoise_2\tortoise.cpp\build\ggml\src\ggml.vcxproj]
     Creating library C:/Users/John/tortoise_2/tortoise.cpp/build/ggml/src/Debug/ggml.lib and object C:/Users/John/tort
  oise_2/tortoise.cpp/build/ggml/src/Debug/ggml.exp
  ggml.vcxproj -> C:\Users\John\tortoise_2\tortoise.cpp\build\bin\Debug\ggml.dll
  Building Custom Rule C:/Users/John/tortoise_2/tortoise.cpp/CMakeLists.txt
  main.cpp
  
  ...
  
  common.cpp
  Generating Code...
  tortoise.vcxproj -> C:\Users\John\tortoise_2\tortoise.cpp\build\tortoise.exe
  Building Custom Rule C:/Users/John/tortoise_2/tortoise.cpp/CMakeLists.txt
````


`ggml.dll` is still generated in `\bin\Debug` and `tortoise.exe` still fails to run unless `ggml.dll` is moved to the build directory.

**Reply by slaren** [reply]
I tried adding `set(BUILD_SHARED_LIBS OFF)` to the llama.cpp `CMakeLists.txt` and it works as expected, I am not sure what you are doing differently.

**Reply by balisujohn** [reply]
strange, I am using a temporary fork of ggml but I haven't messed with the build process at all.

**Reply by balisujohn** [reply]
(I will look at llama.cpp's CMakeLists.txt and look for differences)

**Comment by balisujohn** [comment]
I figured it out, it's because I was designing the process for msvc and ggml's CMakeLists.txt overrides BUILD_SHARED_LIBS to the on state if you are using msvc. https://github.com/ggerganov/ggml/blob/e3b3846976c94163f2b3dd128cc959782653edbb/CMakeLists.txt#L28 Maybe worth fixing in ggml, or maybe I should just switch my compilation process to mingw since msvc is proprietary.

**Reply by balisujohn** [reply]
(I am learning so much about the c++ ecosystem lol)

**Reply by slaren** [reply]
This only changes the default, it doesn't stop applications from setting a different value. I tested it with msvc.
